crates/wasi-common/src/virtfs/pipe.rs

```diff
@@ -0,0 +1,412 @@
+            reader,
+        }
+    }
+
+    /// Try to convert this `ReadPipe<R>` back to the underlying `R` type.
+    ///
+    /// This will fail with `Err(self)` if multiple references to the underlying `R` exist.
+    pub fn try_into_inner(mut self) -> std::result::Result<R, Self> {
+        match Rc::try_unwrap(self.reader) {
+            Ok(rc) => Ok(RefCell::into_inner(rc)),
+            Err(reader) => {
+                self.reader = reader;
+                Err(self)
+            }
+        }
+    }
+}
+
+impl From<Vec<u8>> for ReadPipe<io::Cursor<Vec<u8>>> {
+    fn from(r: Vec<u8>) -> Self {
+        Self::new(io::Cursor::new(r))
+    }
+}
+
+impl From<&[u8]> for ReadPipe<io::Cursor<Vec<u8>>> {
+    fn from(r: &[u8]) -> Self {
+        Self::from(r.to_vec())
+    }
+}
+
+impl From<String> for ReadPipe<io::Cursor<String>> {
+    fn from(r: String) -> Self {
+        Self::new(io::Cursor::new(r))
+    }
+}
+
+impl From<&str> for ReadPipe<io::Cursor<String>> {
+    fn from(r: &str) -> Self {
+        Self::from(r.to_string())
+    }
+}
+
+impl<R: Read + Any> Handle for ReadPipe<R> {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn try_clone(&self) -> io::Result<Box<dyn Handle>> {
+        Ok(Box::new(Self {
+            rights: self.rights.clone(),
+            reader: self.reader.clone(),
+        }))
+    }
+
+    fn get_file_type(&self) -> types::Filetype {
+        types::Filetype::Unknown
+    }
+
+    fn get_rights(&self) -> HandleRights {
+        self.rights.get()
+    }
+
+    fn set_rights(&self, rights: HandleRights) {
+        self.rights.set(rights)
+    }
+
+    fn advise(
+        &self,
+        _advice: types::Advice,
+        _offset: types::Filesize,
+        _len: types::Filesize,
+    ) -> Result<()> {
+        Err(Errno::Spipe)
+    }
+
+    fn allocate(&self, _offset: types::Filesize, _len: types::Filesize) -> Result<()> {
+        Err(Errno::Spipe)
+    }
+
+    fn fdstat_set_flags(&self, _fdflags: types::Fdflags) -> Result<()> {
+        // do nothing for now
+        Ok(())
+    }
+
+    fn filestat_get(&self) -> Result<types::Filestat> {
+        let stat = types::Filestat {
+            dev: 0,
+            ino: 0,
+            nlink: 0,
+            size: 0,
+            atim: 0,
+            ctim: 0,
+            mtim: 0,
+            filetype: self.get_file_type(),
+        };
+        Ok(stat)
+    }
+
+    fn filestat_set_size(&self, _st_size: types::Filesize) -> Result<()> {
+        Err(Errno::Spipe)
+    }
+
+    fn preadv(&self, buf: &mut [io::IoSliceMut], offset: types::Filesize) -> Result<usize> {
+        if offset != 0 {
+            return Err(Errno::Spipe);
+        }
+        Ok(self.reader.borrow_mut().read_vectored(buf)?)
+    }
+
+    fn seek(&self, _offset: io::SeekFrom) -> Result<types::Filesize> {
+        Err(Errno::Spipe)
+    }
+
+    fn read_vectored(&self, iovs: &mut [io::IoSliceMut]) -> Result<usize> {
+        trace!("read_vectored(iovs={:?})", iovs);
```

Comment:
tracing should be (optionally) inserted by wiggle, right? In which case, we don't need the implementation code to manually do it.

Reply:
tracing is always inserted by wiggle at the layer below this, and will trace just the guest pointer. however, without a `tracing_subscriber` enabled in some manner, or the `tracing_log` feature enabled, the traces will act as no-ops. This tracing does show more detail, because the guest pointers have been translated into host slices, which wiggle doesn't know about.